internal/seed/seed.go

```diff
@@ -13,18 +17,96 @@ type Seed struct {
+type Status struct {
+	// Reachable is whether a node was able to be reached or not.
+	Reachable bool
+	// CatchingUp whether a node is still trying to keep up with the network.
+	CatchingUp bool
+	// IsLatestBlock is true when the status of the node is caught up to the latest block.
+	// This together with the block.BlockManager can be leveraged to confirm that the nodes are up-to-date on the latest block.
+	// Because the seeding process takes time, the latest block on the start of the seeding process can be a different one from
+	// the end of the seeding process so an absolute value of the latest block is not a good measurement of the node.
+	// Instead, the singleton block.BlockManager must be used and set to the latest block height and if there is a
+	// block.ErrBlockTooLow when setting the height, it means the node that we queried after was actually not on the latest block yet
+	// and is removed from the seed temporarily.
+	IsLatestBlock bool
 }
 
 type Apis struct {
-	RPC  []Provider `json:"rpc"`
-	Rest []Provider `json:"rest"`
-	GRPC []Provider `json:"grpc"`
+	RPC  []Node `json:"rpc"`
+	Rest []Node `json:"rest"`
+	GRPC []Node `json:"grpc"`
+}
+
+// Seeder represents a seeder process responsible for updating Seed listeners on new changes to the node Seed.
+type Seeder struct {
+	cfg       Config
+	listeners []chan<- Seed
+	init      sync.Once
+	log       *slog.Logger
+	rpcProbe  Probe
+	restProbe Probe
+	grpcProbe Probe
+}
+
+// New creates a Seeder instance.
+// It takes an arbitrary number of listeners that will receive updated Seed structures.
+func New(cfg Config, log *slog.Logger, listeners ...chan<- Seed) *Seeder {
+	return &Seeder{
+		cfg:       cfg,
+		listeners: listeners,
+		log:       log,
+		rpcProbe:  ProbeFunc(RPCProbe),
+		restProbe: ProbeFunc(RESTProbe),
+		grpcProbe: ProbeFunc(GRPCProbe),
+	}
+}
+
+// Start executes a single goroutine to fetch and update seed listeners every Config.SeedRefreshInterval.
+// It is only executed once for every Seeder instance.
+func (s *Seeder) Start(ctx context.Context) {
+	s.log.Info("starting updater")
+	s.init.Do(func() {
+		go func() {
+			t := time.NewTicker(s.cfg.SeedRefreshInterval)
+			defer t.Stop()
+			for {
+				select {
+				case <-t.C:
+					s.fetchAndUpdate(ctx)
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
+		s.fetchAndUpdate(ctx)
+	})
+}
+
+func (s *Seeder) fetchAndUpdate(ctx context.Context) {
+	s.log.Info("fetching seed list")
+	result, err := s.fetch(ctx, s.log, s.cfg.SeedURL)
+	if err != nil {
+		s.log.Error("could not get initial seed list", "err", err)
+		return
+	}
+	if result.ChainID != s.cfg.ChainID {
+		s.log.Error("chain ID is different than expected", "got", result.ChainID, "expected", s.cfg.ChainID)
+		return
+	}
+	for _, ch := range s.listeners {
+		ch <- result
+	}
 }
 
-func fetch(url string) (Seed, error) {
+func (s *Seeder) fetch(ctx context.Context, log *slog.Logger, url string) (Seed, error) {
```

Comment:
@troian I have an [issue](#17) to make logger propagate through the context. I'll not do that in this PR to keep it simpler to review.